### PR TITLE
doc: describe Makefile targets & add reST primer

### PR DESCRIPTION
**Pull request by return42**
With the aim to simplify development cycles, started with PR #1756 a Makefile
based boilerplate was added.  This patch adds the missing developer
documentation.

For a preview use my mirror (don't bookmark):

- [Makefile targets](https://return42.github.io/searx/dev/makefile.html)
- [reST primer](https://return42.github.io/searx/dev/reST.html)
- [Sphinx & Jinja templating](https://return42.github.io/searx/dev/reST.html#templating)
- [Engines](https://return42.github.io/searx/admin/engines.html)
- [Plugins](https://return42.github.io/searx/admin/plugins.html)
- [Code](https://return42.github.io/searx/dev/contribution_guide.html#code)
- [Buildhosts](https://return42.github.io/searx/admin/buildhosts.html)

### CODE REVIEW

**Comment by dalf** [comment]
[This graphviz](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%0A%20%20node%20%5Bstyle%3Dfilled%2Cshape%3Dbox%2C%20fontname%3D%22Arial%22%20fontsize%3D%228%22%5D%3B%0A%20%20edge%20%5Bfontname%3D%22Arial%22%20fontsize%3D%228%22%5D%3B%0A%20%20browser%20%5Blabel%3D%22Browser%22%2Cshape%3DMdiamond%5D%3B%0A%20%20rp%20%5Blabel%3D%22Reverse%20Proxy%22%2C%20href%3D%22url%20to%20configure%20reverse%20proxy%22%5D%3B%0A%20%20filtron%20%5Blabel%3D%22Filtron%22%2C%20href%3D%22https%3A%2F%2Fgithub.com%2Fasciimoo%2Ffiltron%22%5D%3B%0A%20%20morty%20%5Blabel%3D%22Morty%22%2C%20href%3D%22https%3A%2F%2Fgithub.com%2Fasciimoo%2Fmorty%22%5D%3B%0A%20%20static%20%5Blabel%3D%22Static%20files%22%2C%20href%3D%22url%20to%20configure%20static%20files%22%5D%3B%0A%20%20uwsgi%20%5Blabel%3D%22uwsgi%22%2C%20href%3D%22url%20to%20configure%20uwsgi%22%5D%0A%20%20searx1%20%5Blabel%3D%22Searx%20%231%22%5D%3B%0A%20%20searx2%20%5Blabel%3D%22Searx%20%232%22%5D%3B%0A%20%20searx3%20%5Blabel%3D%22Searx%20%233%22%5D%3B%0A%20%20searx4%20%5Blabel%3D%22Searx%20%234%22%5D%3B%0A%0A%20%20browser%20-%3E%20rp%20%5Blabel%3D%22HTTPS%22%5D%0A%20%20subgraph%20cluster_searx%20%7B%0A%20%20%20%20%20%20label%20%3D%20%22Searx%20instance%22%20fontname%3D%22Arial%22%20fontsize%3D%2212%22%3B%0A%20%20%20%20%20%20%7B%20rank%3Dsame%3B%20static%20rp%20%7D%3B%0A%20%20%20%20%20%20rp%20-%3E%20morty%20%5Blabel%3D%22optional%5Cnimages%20and%20HTML%20pages%20proxy%22%5D%3B%0A%20%20%20%20%20%20rp%20-%3E%20static%20%5Blabel%3D%22optional%5CnThe%20reverse%20proxy%20serves%20directly%20the%20static%20files%22%5D%3B%0A%20%20%20%20%20%20rp%20-%3E%20filtron%20%5Blabel%3D%22HTTP%22%5D%3B%0A%20%20%20%20%20%20filtron%20-%3E%20uwsgi%20%5Blabel%3D%22HTTP%22%5D%0A%20%20%20%20%20%20uwsgi%20-%3E%20searx1%3B%0A%20%20%20%20%20%20uwsgi%20-%3E%20searx2%3B%0A%20%20%20%20%20%20uwsgi%20-%3E%20searx3%3B%0A%20%20%20%20%20%20uwsgi%20-%3E%20searx4%3B%0A%20%20%7D%0A%0A%7D) may help to understand the architecture (If this is the right one).

```graphviz
digraph G {

  node [style=filled,shape=box, fontname="Arial" fontsize="8"];
  edge [fontname="Arial" fontsize="8"];
  browser [label="Browser",shape=Mdiamond];
  rp [label="Reverse Proxy", href="url to configure reverse proxy"];
  filtron [label="Filtron", href="https://github.com/asciimoo/filtron"];
  morty [label="Morty", href="https://github.com/asciimoo/morty"];
  static [label="Static files", href="url to configure static files"];
  uwsgi [label="uwsgi", href="url to configure uwsgi"]
  searx1 [label="Searx #1"];
  searx2 [label="Searx #2"];
  searx3 [label="Searx #3"];
  searx4 [label="Searx #4"];

  browser -> rp [label="HTTPS"]
  subgraph cluster_searx {
      label = "Searx instance" fontname="Arial" fontsize="12";
      { rank=same; static rp };
      rp -> morty [label="optional\nimages and HTML pages proxy"];
      rp -> static [label="optional\nThe reverse proxy serves directly the static files"];
      rp -> filtron [label="HTTP"];
      filtron -> uwsgi [label="HTTP"]
      uwsgi -> searx1;
      uwsgi -> searx2;
      uwsgi -> searx3;
      uwsgi -> searx4;
  }

}
```

**Comment by return42** [comment]
> This graphviz may help to understand the architecture (If this is the right one).

Good Point, I will add it to the docs .. is there any origin for, or is the link already the origin?

**Comment by dalf** [comment]
I'm not sure to understand what you mean by "origin".

About the links between the components: 
* uwsgi: it is documented here https://asciimoo.github.io/searx/dev/install/installation.html but some people doesn't use it (according to the various installation procedure that can be found).
* about the "static files": I've setup this few years ago, it doesn't seems to be documented actually.

**Comment by return42** [comment]
> I'm not sure to understand what you mean by "origin".

If you copied this image from *another place* it's origin is the "other place" and we have to ask about the licenses.  But I guess it was made by yourself (and the link is the only origin)?

BTW:  I tested the your DOT markup, but I got problems when it renders the selected font Arial .. which is proprietary .. and not installed on most (build) linux servers.  To me it seems, that the fallback font has different sizes and the calculated boxes sizes (SVG) are smaller then the lines of text (overwrite effects).  Sorry for the bad description .. I haven't had time to dig deeper into digraph, and fonts at built time and fonts which are used in the viewer on client site, when the SVG is rendered and an explicit font (Arial) is selected in the SVG (xml) but not installed on the system.

**Comment by dalf** [comment]
* Yes I made it myself.
* ```Liberation Sans``` or ```DejaVu Sans``` are okay.

Licenses:
* ```DejaVu Sans```: see the [Wikipedia page](https://en.wikipedia.org/wiki/DejaVu_fonts)
* ```Liberation```: [SIL Open Font License](https://en.wikipedia.org/wiki/SIL_Open_Font_License) according to [Wikipedia](https://en.wikipedia.org/wiki/Liberation_fonts )

**Comment by return42** [comment]
@dalf: I added your DOT graph (see aa3b0265).  I got best result with fontsize leafed unset and font family 'Sans'.  If needed, we can change this later, when PR is merge upstream.

preview (don't bookmark): https://return42.github.io/searx/admin/architecture.html

**Comment by dalf** [comment]
I'm not sure this is the right place. In my opinion, most of the wiki should be moved to the reST documentation:
* settings.yml : https://github.com/asciimoo/searx/wiki/settings.yml : description of the general settings.
* docker: https://github.com/asciimoo/searx/wiki/Docker-developement-environment : to merge/remove most probably
* engines: https://github.com/asciimoo/searx/wiki/existing-engines : to remove most probably, since it is generate automatically with this PR.

About the reverse proxy configuration / installation:
* https://github.com/asciimoo/searx/wiki/How-to-create-and-configure-SSL
* https://github.com/asciimoo/searx/wiki/Searx-with-haproxy
* https://github.com/asciimoo/searx/wiki/Web-server-hardening-(Apache) : to remove (?) see below. 
* https://github.com/asciimoo/searx/wiki/Installation-on-RHEL-7---CentOS-7
* https://github.com/asciimoo/searx/wiki/Installation : to remove

A way to deal with the reverse proxy configurations (if possible):
* define some variables (from [this](https://github.com/searx/searx-docker/blob/f2e6ca4a610d9ccf66d4e16ac210bdb1bd4fff0c/Caddyfile#L19) or somewhere else)
* generate the Apache, nginx, caddy, haproxy configurations using some templates, useful especially for the HTTP headers (Content-Security-Policy header etc...). 

It would avoid different versions of the configuration.

Perhaps, each reverse proxy should have it's own page in the reST documentation.

Anyway, most probably this PR must be merged in the master branch before dealing with the wiki.

**Comment by return42** [comment]
> I'm not sure this is the right place. In my opinion, most of the wiki should be moved to the reST documentation

It was one of my intentions, good to hear that other think same.

> Anyway, most probably this PR must be merged in the master branch before dealing with the wiki.

@dalf do you like to open a issue with a copy of your comment above / mark it as 'doc'' and we an pick up it after this PR is merged upstream.

**Comment by return42** [comment]
Is there anyone who has time to review / merge this PR upstream? (cant merge my own PR)

BTW: we need to update gh-pages, this can be done with:

    make clean gh-pages

Should I do? .. or who maintains gh-pages branch?

File: requirements-dev.txt
Comment (dalf) [inline requirements-dev.txt:14]: It would be better to depend on a release rather than the last commit.

Comment (return42) [reply requirements-dev.txt:14]: As long, as we do not have patched releases, searx itself is always updated to last commit on searx/master.

I understand your concerns and if you want, I can changed the requirement to a fixed one.  But let me also hint, that I am the maintainer of linuxdoc.  We also do not maintain releases at linuxdoc and recommend to regularly update instances to topmost commit master.

Should I changed to a fixed version or is it OK for you to go with the linuxdoc master branch?

**Comment by dalf** [comment]
Some notes: 
* reST.html includes a math rendering example, so Sphinx loads mathjax from the Cloudflare CDN. I think all resources should be loaded from the host, and not use a CDN. So in my opinion, the math equation example should be removed (see https://github.com/sphinx-doc/sphinx/pull/5536 ). If it is really needed we can look at https://docs.mathjax.org/en/latest/web/hosting.html#hosting-your-own-copy-of-mathjax (way more complicated for sure).
* Personal taste: dev/reST.html : The zoom effect on grouped tab makes the reading difficult.

**Comment by return42** [comment]
> so Sphinx loads mathjax from the Cloudflare CDN

ups, I haven't been aware of.   I Have to look for a solution that does not load content from CDNs!

> The zoom effect on grouped tab makes the reading difficult.

I use it for a long time, but I also doubt  since that time that this is the solution i wanted.

I looking for a better solution, do you have any preferences?

And: thanks for your deep inspection!

**Comment by dalf** [comment]
> I looking for a better solution, do you have any preferences?

I prefer without the animation effect, so removing the lines 75, 76 and 82 would be enough : 
https://github.com/asciimoo/searx/pull/1776/files#diff-7e4e00712fa761cac6e56c7591bffb0cR75-R76

**Comment by return42** [comment]
> so Sphinx loads mathjax from the Cloudflare CDN

this is fixed with d1892b2 / the math are rendered as images

> The zoom effect on grouped tab makes the reading difficult.

this is fixed with b91e07b the rst-example is rendered slightly more discreet 

preview (don't bookmark): https://return42.github.io/searx/admin/buildhosts.html

@dalf, do you have time to make a review of this PR on github?

**Review by dalf — APPROVED** [review]
It's all good, except:
* the column {{mod.suspend_end_time}} to remove (in engines.rst)
* the href links to remove in arch_public.dot

Anyway I don't think the purpose of this PR is to rewrite all the documentation in one shot, so let's go. 
Thank you !

File: docs/admin/arch_public.dot
Comment (return42) [reply docs/admin/arch_public.dot:11]: You have to open the svg image (right click "View Image) .. or directly open https://url-too/arch_public.svg .. This is my UX from FFox at least.

File: docs/dev/makefile.rst
Comment (return42) [reply docs/dev/makefile.rst:1]: this is not correct, you can build without graphviz /...

> if a third party converter is missed, the build process will spit out a message about and continues with a lower quality in the output. Even if the output is e.g. a raw SVG or DOT, in some use cases it might be a solution that fits

see [Scalable figure and image handling](https://return42.github.io/linuxdoc/linuxdoc-howto/kfigure.html)

I guess most developer or admins will never need to build with the full buildchain installed. Thats why I added the [buildhosts article](https://return42.github.io/searx/admin/buildhosts.html) article.

BTW: the math to svg conversion process is handled similar ..  it spits out warnings (see buiildhost article)  if a tool is missed, but the build process will continue (its a warning not an error).

File: docs/admin/engines.rst
Comment (return42) [reply docs/admin/engines.rst:66]: Thanks for pointing out .. I fix it afterwards.

**Comment by ahmadpanahhamid** [comment]
The goal is to identify those who work hard and those who unjustly publicize their efforts and earn a living.
The goal is that some people create a work, but there are people who destroy their rights.
There is always hope for the rights of a human being, although all of us human beings make mistakes, please respect the rights as well.

**Pull request merged**
